diff --git a/appengine/tools/devappserver2/api_server.py b/appengine/tools/devappserver2/api_server.py
index 5ccf415..76b411e 100644
--- a/appengine/tools/devappserver2/api_server.py
+++ b/appengine/tools/devappserver2/api_server.py
@@ -58,6 +58,7 @@ from google.appengine.api.system import system_stub
 from google.appengine.api.xmpp import xmpp_service_stub
 from google.appengine.datastore import datastore_sqlite_stub
 from google.appengine.datastore import datastore_stub_util
+from google.appengine.datastore import datastore_mongodb_stub
 
 from google.appengine.api import apiproxy_stub_map
 from google.appengine.api import datastore
@@ -304,13 +305,10 @@ def setup_stubs(
       'channel',
       channel_service_stub.ChannelServiceStub(request_data=request_data))
 
-  datastore_stub = datastore_sqlite_stub.DatastoreSqliteStub(
+  datastore_stub = datastore_mongodb_stub.DatastoreMongoDBStub(
       app_id,
-      datastore_path,
-      datastore_require_indexes,
-      trusted,
-      root_path=application_root,
-      auto_id_policy=datastore_auto_id_policy)
+      require_indexes=datastore_require_indexes,
+      root_path=application_root)
 
   datastore_stub.SetConsistencyPolicy(datastore_consistency)
 
